File: Advent of Code/2020/Day 15/Day 15 Part 2.py
import time

start_time = time.time()

example1 = '0,3,6'.split(',')
example2 = '1,3,2'.split(',')
example3 = '2,1,3'.split(',')
example4 = '1,2,3'.split(',')
example5 = '2,3,1'.split(',')
example6 = '3,2,1'.split(',')
example7 = '3,1,2'.split(',')
input_numbers = '2,0,6,12,1,3'.split(',')

numbers = input_numbers

seen = {}
l = len(numbers)
current = numbers[l-1]
f = 30000000

def addtodict(key, val):
    key = str(key)
    seen[key] = { "last": val }

for i in range(0, l - 1):
    addtodict(numbers[i], i + 1)

for j in range(l, f):
    if current in seen:
        new = str(j - seen[current]['last'])
    else:
        new = "0"
    addtodict(current, j)
    current = new

print('Answer is ' + current)
print("--- %s seconds ---" % (time.time() - start_time))

File: Advent of Code/2020/Day 15/Day 15 Part 2.R
library(readr)
library(tidyverse)

start_time <- Sys.time()

numbers <-read_lines("example.txt") %>% str_split(",") %>% unlist()
l <- length(numbers)
seen <- list(1,2,3)
names(seen) <- numbers[1:(l-1)]
current <- numbers[l]
f <- 2020

for (i in l:(f-1)){
  if(current %in% names(seen)){
    new <- as.character(i - seen[[current]])
    seen[current] <- i
  } else {
    new <- "0"
    seen[current] <- i
  }
  current <- new
  if(i %% 100000 == 0) {print(i)}
}

print(paste0("Answer is ", current))

print(Sys.time() - start_time)
